core/lib/dal/src/blocks_dal.rs

```diff
@@ -2334,6 +2355,80 @@ impl BlocksDal<'_, '_> {
 
         Ok(results.into_iter().map(L::from).collect())
     }
+
+    pub async fn has_last_l2_block_bloom(&mut self) -> DalResult<bool> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                (logs_bloom IS NOT NULL) AS "logs_bloom_not_null!"
+            FROM
+                miniblocks
+            ORDER BY
+                number DESC
+            LIMIT
+                1
+            "#,
+        )
+        .instrument("has_last_l2_block_bloom")
+        .fetch_optional(self.storage)
+        .await?;
+
+        Ok(row.map(|row| row.logs_bloom_not_null).unwrap_or(false))
+    }
+
+    pub async fn get_max_l2_block_without_bloom(&mut self) -> DalResult<Option<L2BlockNumber>> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                MAX(number) AS "max?"
+            FROM
+                miniblocks
+            WHERE
+                logs_bloom IS NULL
+            "#,
+        )
+        .instrument("get_max_l2_block_without_bloom")
+        .fetch_one(self.storage)
+        .await?;
+
+        Ok(row.max.map(|n| L2BlockNumber(n as u32)))
+    }
+
+    pub async fn range_update_logs_bloom(
+        &mut self,
+        from_l2_block: L2BlockNumber,
+        blooms: &[Bloom],
+    ) -> DalResult<()> {
+        let to_l2_block = from_l2_block + (blooms.len() - 1) as u32;
```

Comment:
Nit: Not sure whether it can be triggered, but adding parentheses means an integer underflow if `blooms` are empty. To be fair, it would be corrected during addition; the only worrying case is if `blooms` are empty and `from_l2_block == L2BlockNumber(0)`. Could return early if `blooms` are empty to avoid these cases.